kernel: k_sleep doesn't handle relative or absolute timeouts >INT_MAX

**Describe the bug**

The primary sleep function `z_tick_sleep` in `kernel/sched.c` does not respect the type of `k_ticks_t`, instead using `int32_t`.
There are two main consequences of this:

1. Sleeps of tick duration greater than `INT_MAX` will not behave properly.
2. Absolute timeouts fail after `INT_MAX` ticks (18 hours on nRF).

In the second case, timout.ticks will have some value less than `INT_MIN`. This becomes a postive number near `INT_MAX` after being cast to `int32_t`. The `Z_TICK_ABS` check in `z_add_timeout` fails, resulting in a sleep of approximately 18hrs being scheduled, instead of ~1 second.

**To Reproduce**

Initialize `curr_tick` in `kernel\timeout.c` to just before rollover:
```
static uint64_t curr_tick = INT32_MAX - 320000;
``` 
Modify the main loop of `samples\basic\blinky` to the following:
```
	int64_t timeout = k_uptime_get() + 1000;

	while (1) {
		k_sleep(K_TIMEOUT_ABS_MS(timeout));
		timeout += 1000;

		gpio_pin_set(dev, PIN, (int)led_is_on);
		led_is_on = !led_is_on;
	}
```
On NRF boards, the LED will stop blinking after ~10 seconds.

**Expected behavior**
Absolute timeouts (and large relative timeouts) should delay for the specified duration.

**Impact**
High priority/showstopper, I use absolute timeouts to mitigate drift in our main loop.
This bug means the watchdog is rebooting my application every 18 hours.

**Logs and console output**
By adding `printk("Ticks: %lld %lld\n", ticks, Z_TICK_ABS(ticks));` to `z_add_timeout`
```
Ticks: -2147324823 2147324821
Ticks: -2147357591 2147357589
Ticks: -2147390359 2147390357
Ticks: -2147423127 2147423125
Ticks: -2147455895 2147455893
Ticks: 2147478633 -2147478635
```

**Environment (please complete the following information):**
- Zephyr 2.4

Adding `#pragma GCC diagnostic error "-Wconversion"` to the top of `sched.c` does produce (among a multitude of others):
```
C:/Users/Jordan/code/project/zephyr/kernel/sched.c: In function 'z_impl_k_sleep':
C:/Users/Jordan/code/project/zephyr/kernel/sched.c:1265:23: error: conversion from 'k_ticks_t' {aka 'long long int'} to 'int32_t' {aka 'int'} may change value [-Werror=conversion]
 1265 |  ticks = z_tick_sleep(ticks);
```
Based on `compiler_flags.cmake`, this option is only enabled for warning levels 2 and 3.
Unfortunately there are so many warnings at these levels that they are functionally useless.

If the build system supported it, I feel like it would be useful for the core kernel functionality to be compiled with the highest possible warning levels, even if device drivers etc aren't. That would help pick up on bugs like this one.

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -1240,7 +1240,7 @@ static inline void z_vrfy_k_yield(void)
 static int32_t z_tick_sleep(k_ticks_t ticks)
 {
 #ifdef CONFIG_MULTITHREADING
-	uint32_t expected_wakeup_time;
+	uint32_t expected_wakeup_ticks;
 
 	__ASSERT(!arch_is_in_isr(), "");
 
@@ -1266,7 +1266,7 @@ static int32_t z_tick_sleep(k_ticks_t ticks)
 	timeout = Z_TIMEOUT_TICKS(ticks);
 #endif
 
-	expected_wakeup_time = ticks + z_tick_get_32();
+	expected_wakeup_ticks = ticks + z_tick_get_32();
 
 	k_spinlock_key_t key = k_spin_lock(&sched_spinlock);
 
@@ -1281,7 +1281,7 @@ static int32_t z_tick_sleep(k_ticks_t ticks)
 
 	__ASSERT(!z_is_thread_state_set(_current, _THREAD_SUSPENDED), "");
 
-	ticks = (k_ticks_t)expected_wakeup_time - z_tick_get_32();
+	ticks = (k_ticks_t)expected_wakeup_ticks - z_tick_get_32();
 	if (ticks > 0) {
 		return ticks;
 	}

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -1237,14 +1237,19 @@ static inline void z_vrfy_k_yield(void)
 #include <syscalls/k_yield_mrsh.c>
 #endif
 
-static int32_t z_tick_sleep(int32_t ticks)
+static int32_t z_tick_sleep(k_ticks_t ticks)
 {
 #ifdef CONFIG_MULTITHREADING
 	uint32_t expected_wakeup_time;
 
 	__ASSERT(!arch_is_in_isr(), "");
 
-	LOG_DBG("thread %p for %d ticks", _current, ticks);
+#ifndef CONFIG_TIMEOUT_64BIT
+	/* LOG subsys does not handle 64-bit values
+	 * https://github.com/zephyrproject-rtos/zephyr/issues/26246
+	 */
+	LOG_DBG("thread %p for %u ticks", _current, ticks);
+#endif
 
 	/* wait of 0 ms is treated as a 'yield' */
 	if (ticks == 0) {
@@ -1258,7 +1263,7 @@ static int32_t z_tick_sleep(int32_t ticks)
 	timeout = Z_TIMEOUT_TICKS(ticks);
 #else
 	ticks += _TICK_ALIGN;
-	timeout = (k_ticks_t) ticks;
+	timeout = Z_TIMEOUT_TICKS(ticks);
 #endif
 
 	expected_wakeup_time = ticks + z_tick_get_32();
@@ -1276,7 +1281,7 @@ static int32_t z_tick_sleep(int32_t ticks)
 
 	__ASSERT(!z_is_thread_state_set(_current, _THREAD_SUSPENDED), "");
 
-	ticks = expected_wakeup_time - z_tick_get_32();
+	ticks = (k_ticks_t)expected_wakeup_time - z_tick_get_32();
 	if (ticks > 0) {
 		return ticks;
 	}

